src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs

```diff
@@ -0,0 +1,1404 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// Imported from https://github.com/dotnet/runtime/blob/c7804d5b3c8bd32e35cbb674e8f275fcaf754d93/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs
+
+#define TARGET_64BIT // Use runtime detection of 64bits target
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.Serialization;
+using Uno.Foundation;
+
+namespace Windows.UI.Xaml
+{
+	/// <summary>
+	/// Specialized Dictionary for ResourceDictionary values backing, using <see cref="ResourceKey"/> for the dictionary key.
+	/// </summary>
+	internal class SpecializedResourceDictionary
+	{
+		/// <summary>
+		/// Represents a key for the source dictionary
+		/// </summary>
+		public readonly struct ResourceKey
+		{
+			public readonly string Key;
+			public readonly bool IsType;
+			public readonly uint HashCode;
+
+			public static ResourceKey Empty { get; } = new ResourceKey(false);
+
+			public bool IsEmpty => Key == null;
+
+			private ResourceKey(bool dummy)
+			{
+				Key = null;
+				IsType = false;
+				HashCode = 0;
+			}
+
+			/// <summary>
+			/// Builds a ResourceKey based on an unknown object
+			/// </summary>
+			/// <param name="key">The original key to use</param>
+			public ResourceKey(object key)
+			{
+				if (key is string s)
+				{
+					Key = s;
+					IsType = false;
+					HashCode = (uint)(s.GetHashCode() ^ IsType.GetHashCode());
+				}
+				else if (key is Type t)
+				{
+					Key = t.ToString();
+					IsType = true;
+					HashCode = (uint)(t.GetHashCode() ^ IsType.GetHashCode());
+				}
+				else
+				{
+					Key = key.ToString();
```

Comment:
I feel like there could be the potential for duplicate keys in corner cases involving non-string/non-type objects used as keys (ie two different objects returning the same `ToString()` value). (Disclaimer: I don't even know if UWP properly supports non-string/non-type objects as ResourceDictionary keys.)